File: crates/session-host/src/link.rs
//! The client end of the pipe: connecting, verifying, and one connection's
//! traffic.
//!
//! This lives in the session host crate rather than in the Worker for a
//! practical reason as much as a tidy one. The Worker cannot be
//! cross-compiled for Windows on a Mac — one of its transitive dependencies
//! needs the Windows SDK headers (`docs/research/m0-executor-probes.md`) — so
//! Windows code that sits there is written blind. This crate *can* be
//! cross-compiled, so everything Win32-shaped is put here, where
//! `cargo check --target x86_64-pc-windows-msvc` is at least an opinion.
//!
//! What is left in the Worker is the mapping to its own backend trait, and
//! nothing that touches a handle.

#![cfg(windows)]

use std::{
    collections::HashMap,
    io,
    path::Path,
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
    time::Duration,
};

use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    net::windows::named_pipe::{ClientOptions, NamedPipeClient},
    sync::{Mutex, oneshot},
};

use crate::{
    PROTOCOL_MAJOR,
    client::{Delivery, Greeting, OutputTracker, accept_welcome},
    pipe,
    protocol::{ClientMessage, Frame, FrameDecoder, FrameKind, HostMessage, encode_json},
};

/// How long a request may wait before the caller is told the host is
/// unreachable, rather than being left on a future that will never resolve.
pub const REQUEST_TIMEOUT: Duration = Duration::from_secs(10);

/// How long to keep trying after starting the host.
pub const START_TIMEOUT: Duration = Duration::from_secs(5);

/// The file name the Worker looks for beside itself.
pub const HOST_BINARY: &str = "armadra-session-host.exe";

/// What a connection hands back to whoever opened it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LinkEvent {
    /// Replay of what happened before this attachment existed.
    Snapshot {
        generation: u64,
        payload: Vec<u8>,
    },
    /// The replay is complete; everything after this is live.
    SnapshotEnd {
        generation: u64,
    },
    Output {
        generation: u64,
        payload: Vec<u8>,
    },
    /// Frames were lost. The screen cannot be repaired by writing what
    /// arrived, so the caller has to attach again for a fresh replay.
    Gap {
        generation: u64,
        missing: u64,
    },
    Exit {
        session_key: String,
        generation: u64,
        exit_code: Option<i64>,
    },
    Stale {
        session_key: String,
        generation: u64,
        current: u64,
    },
    Warning {
        session_key: String,
        message: String,
    },
    Bye {
        reason: String,
        drain: bool,
    },
    /// The connection ended, for any reason.
    Closed,
}

/// Where a connection's events go. Called from the reader task, so it must not
/// block.
pub type EventSink = Arc<dyn Fn(LinkEvent) + Send + Sync>;

/// One connected pipe with a task reading it.
pub struct Link {
    outgoing: tokio::sync::mpsc::Sender<Frame>,
    pending: Arc<Mutex<HashMap<u64, oneshot::Sender<HostMessage>>>>,
    welcome: Mutex<Option<oneshot::Receiver<HostMessage>>>,
    tracker: Arc<Mutex<Option<OutputTracker>>>,
    closed: Arc<AtomicBool>,
}

impl Link {
    /// Connects to `endpoint` and starts serving it.
    ///
    /// The pipe name is derived and therefore predictable, so being connected
    /// proves nothing on its own: the server's identity is checked on this
    /// exact handle before any request is sent or any output is read.
    pub async fn connect(endpoint: &str, events: EventSink) -> io::Result<Arc<Self>> {
        let pipe = ClientOptions::new().open(endpoint)?;
        verify_server(&pipe)?;
        Ok(Arc::new(Self::serve(pipe, events)))
    }

    fn serve(pipe: NamedPipeClient, events: EventSink) -> Self {
        let (outgoing, mut queue) = tokio::sync::mpsc::channel::<Frame>(256);
        let pending: Arc<Mutex<HashMap<u64, oneshot::Sender<HostMessage>>>> =
            Arc::new(Mutex::new(HashMap::new()));
        let tracker: Arc<Mutex<Option<OutputTracker>>> = Arc::new(Mutex::new(None));
        let closed = Arc::new(AtomicBool::new(false));
        let (welcome_sender, welcome) = oneshot::channel();

        let (mut reader, mut writer) = tokio::io::split(pipe);
        let write_closed = closed.clone();
        tokio::spawn(async move {
            while let Some(frame) = queue.recv().await {
                let Ok(bytes) = frame.encode() else { continue };
                if writer.write_all(&bytes).await.is_err() {
                    break;
                }
            }
            write_closed.store(true, Ordering::SeqCst);
        });

        let read_pending = pending.clone();
        let read_tracker = tracker.clone();
        let read_closed = closed.clone();
        tokio::spawn(async move {
            let mut decoder = FrameDecoder::new();
            let mut buffer = vec![0u8; 64 * 1024];
            let mut welcome_sender = Some(welcome_sender);
            'outer: loop {
                match reader.read(&mut buffer).await {
                    Ok(0) | Err(_) => break,
                    Ok(count) => decoder.push(&buffer[..count]),
                }
                loop {
                    match decoder.next_frame() {
                        Ok(Some(frame)) => {
                            route(
                                frame,
                                &read_pending,
                                &read_tracker,
                                &events,
                                &mut welcome_sender,
                            )
                            .await;
                        }
                        Ok(None) => break,
                        Err(_) => break 'outer,
                    }
                }
            }
            read_closed.store(true, Ordering::SeqCst);
            // Waking every waiter is what keeps a dead host from turning into
            // a set of futures nobody ever resolves.
            read_pending.lock().await.clear();
            events(LinkEvent::Closed);
        });

        Self {
            outgoing,
            pending,
            welcome: Mutex::new(Some(welcome)),
            tracker,
            closed,
        }
    }

    pub fn alive(&self) -> bool {
        !self.closed.load(Ordering::SeqCst)
    }

    /// Closes this connection. For an attached connection that is a detach:
    /// the host keeps the session, only `destroy` ends one.
    pub fn close(&self) {
        self.closed.store(true, Ordering::SeqCst);
        self.outgoing.try_send(Frame::json(Vec::new())).ok();
    }

    pub async fn handshake(&self, client: &str) -> io::Result<Greeting> {
        self.notify(&ClientMessage::Hello {
            protocol: PROTOCOL_MAJOR,
            client: client.to_owned(),
        })
        .await?;
        let receiver = self
            .welcome
            .lock()
            .await
            .take()
            .ok_or_else(|| io::Error::other("this connection has already been greeted"))?;
        let message = tokio::time::timeout(REQUEST_TIMEOUT, receiver)
            .await
            .map_err(|_| io::Error::other("the session host did not answer hello"))?
            .map_err(|_| io::Error::other("the session host closed during hello"))?;
        accept_welcome(&message, PROTOCOL_MAJOR).map_err(io::Error::other)
    }

    /// Sends without waiting for an answer.
    pub async fn notify(&self, message: &ClientMessage) -> io::Result<()> {
        let frame = encode_json(message).map_err(|error| io::Error::other(error.to_string()))?;
        self.outgoing
            .send(frame)
            .await
            .map_err(|_| io::Error::other("the session host connection is closed"))
    }

    /// Sends and waits for the reply carrying `id`.
    pub async fn request(&self, id: u64, message: &ClientMessage) -> io::Result<HostMessage> {
        let (sender, receiver) = oneshot::channel();
        self.pending.lock().await.insert(id, sender);
        if let Err(error) = self.notify(message).await {
            self.pending.lock().await.remove(&id);
            return Err(error);
        }
        match tokio::time::timeout(REQUEST_TIMEOUT, receiver).await {
            Ok(Ok(message)) => Ok(message),
            Ok(Err(_)) => Err(io::Error::other("the session host closed before answering")),
            Err(_) => {
                self.pending.lock().await.remove(&id);
                Err(io::Error::other("the session host timed out"))
            }
        }
    }

    /// Starts checking output frames against `generation`. Called before the
    /// `attach` request goes out, so no frame can arrive unchecked.
    pub async fn expect_output(&self, generation: u64) {
        *self.tracker.lock().await = Some(OutputTracker::new(generation));
    }
}

async fn route(
    frame: Frame,
    pending: &Arc<Mutex<HashMap<u64, oneshot::Sender<HostMessage>>>>,
    tracker: &Arc<Mutex<Option<OutputTracker>>>,
    events: &EventSink,
    welcome: &mut Option<oneshot::Sender<HostMessage>>,
) {
    match frame.kind {
        // A snapshot is the past being redrawn. It carries no sequence and
        // must not be checked against the live stream's numbering.
        FrameKind::Snapshot => events(LinkEvent::Snapshot {
            generation: frame.generation,
            payload: frame.payload,
        }),
        FrameKind::SnapshotEnd => events(LinkEvent::SnapshotEnd {
            generation: frame.generation,
        }),
        FrameKind::Output => {
            let mut held = tracker.lock().await;
            let Some(state) = held.as_mut() else { return };
            match state.observe(frame.generation, frame.sequence) {
                Delivery::Write => events(LinkEvent::Output {
                    generation: frame.generation,
                    payload: frame.payload,
                }),
                // Belongs to a generation this attachment is not showing.
                Delivery::Wrong => {}
                Delivery::Gap { missing } => {
                    held.take();
                    events(LinkEvent::Gap {
                        generation: frame.generation,
                        missing,
                    });
                }
            }
        }
        FrameKind::Json => {
            let Ok(message) = serde_json::from_slice::<HostMessage>(&frame.payload) else {
                return;
            };
            match message {
                HostMessage::Welcome { .. } => {
                    if let Some(sender) = welcome.take() {
                        let _ = sender.send(message);
                    }
                }
                HostMessage::Ok { id, .. } | HostMessage::Error { id, .. } => {
                    if let Some(sender) = pending.lock().await.remove(&id) {
                        let _ = sender.send(message);
                    }
                }
                HostMessage::Exit {
                    session_key,
                    generation,
                    exit_code,
                } => events(LinkEvent::Exit {
                    session_key,
                    generation,
                    exit_code,
                }),
                HostMessage::Stale {
                    session_key,
                    generation,
                    current,
                } => {
                    tracker.lock().await.take();
                    events(LinkEvent::Stale {
                        session_key,
                        generation,
                        current,
                    });
                }
                HostMessage::Warning {
                    session_key,
                    message,
                } => events(LinkEvent::Warning {
                    session_key,
                    message,
                }),
                HostMessage::Bye { reason, drain } => events(LinkEvent::Bye { reason, drain }),
            }
        }
    }
}

/* -------------------------------- discovery -------------------------------- */

/// The pipe this user's host of this data directory listens on.
pub fn endpoint_for(data_dir: &str) -> io::Result<String> {
    Ok(pipe::endpoint(&crate::winsec::current_sid()?, data_dir))
}

/// Starts the host detached.
///
/// Detached matters: a host started as an ordinary child would share this
/// process' console and process group, and terminal host design §3 rule 4 is
/// explicit that the process owning the consoles must not be tied to one that
/// exits.
pub fn start_host(executable: &Path, data_dir: &Path) -> io::Result<()> {
    use std::os::windows::process::CommandExt;
    // CREATE_NEW_PROCESS_GROUP | DETACHED_PROCESS | CREATE_NO_WINDOW
    const FLAGS: u32 = 0x0000_0200 | 0x0000_0008 | 0x0800_0000;
    std::process::Command::new(executable)
        .arg(data_dir.as_os_str())
        .creation_flags(FLAGS)
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .spawn()
        .map(|_| ())
}

/// Connects, retrying with backoff until `within` elapses.
pub async fn wait_for_host(
    endpoint: &str,
    events: EventSink,
    within: Duration,
) -> io::Result<Arc<Link>> {
    let deadline = tokio::time::Instant::now() + within;
    let mut wait = Duration::from_millis(50);
    loop {
        match Link::connect(endpoint, events.clone()).await {
            Ok(link) => return Ok(link),
            Err(error) if tokio::time::Instant::now() >= deadline => return Err(error),
            Err(_) => {
                tokio::time::sleep(wait).await;
                wait = (wait * 2).min(Duration::from_millis(500));
            }
        }
    }
}

/// The host binary, beside the caller's own executable.
pub fn resolve_host_binary() -> io::Result<std::path::PathBuf> {
    if let Some(path) = std::env::var_os("ARMADRA_SESSION_HOST_BIN") {
        let path = std::path::PathBuf::from(path);
        if path.is_file() {
            return Ok(path);
        }
        return Err(io::Error::new(
            io::ErrorKind::NotFound,
            format!(
                "ARMADRA_SESSION_HOST_BIN does not point at a file: {}",
                path.display()
            ),
        ));
    }
    let executable = std::env::current_exe()?;
    let sibling = executable
        .parent()
        .ok_or_else(|| io::Error::other("this executable has no directory"))?
        .join(HOST_BINARY);
    if sibling.is_file() {
        return Ok(sibling);
    }
    Err(io::Error::new(
        io::ErrorKind::NotFound,
        format!(
            "could not find {HOST_BINARY} next to {}",
            executable.display()
        ),
    ))
}

/// Checks that the process serving this pipe is one we are willing to talk to.
fn verify_server(pipe: &NamedPipeClient) -> io::Result<()> {
    use std::os::windows::io::AsRawHandle;
    let expected = crate::winsec::current_sid()?;
    let actual = unsafe { crate::winsec::server_sid(pipe.as_raw_handle().cast())? };
    if !pipe::principal_allowed(&expected, &actual) {
        return Err(io::Error::new(
            io::ErrorKind::PermissionDenied,
            "the session host pipe belongs to a different principal",
        ));
    }
    Ok(())
}

File: apps/desktop/src/core/terminal/session-host/link.ts
import { execFileSync, spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { connect, type Socket } from "node:net";
import { dirname, join } from "node:path";
import {
  type ClientMessage,
  FrameDecoder,
  type Frame,
  type Greeting,
  type HelloAuth,
  type HostMessage,
  HOST_BINARY,
  OutputTracker,
  PROTOCOL_MAJOR,
  REQUEST_TIMEOUT_MS,
  START_TIMEOUT_MS,
  acceptWelcome,
  jsonFrame,
  pipeEndpoint,
} from "./protocol";

/**
 * One connected pipe, with a reader draining it.
 *
 * The protocol is in `protocol.ts` and is pure; everything that touches a
 * handle is here, which is exactly the split the Rust side made for the same
 * reason — that code could be cross-compiled and inspected, this code can only
 * be *run*, and no machine in this project's CI is a Windows machine.
 *
 * ## What is verified, and what is not
 *
 * The pipe name is derived and therefore predictable, so being connected
 * proves nothing on its own. The Rust client answered that with
 * `verify_server`: `GetNamedPipeServerProcessId` on the handle, then the
 * server process' SID compared against this user's.
 *
 * **`node:net` cannot do that.** It hands back a `Socket`, not a `HANDLE`,
 * and there is no Node API that reaches the pipe's server identity. Since
 * R6d the check is mutual instead of native: the client proves it can read
 * the `0600` key file under the data directory (`auth.ts`), and the host
 * refuses a connection that cannot. That authenticates the *conversation*
 * rather than the peer process, which against the threat that matters — a
 * different account on a shared machine — is the same thing the DACL bought.
 * The residual risk is enumerated in `auth.ts` rather than left implicit.
 *
 * This direction — client verifying server — is still the weaker one: a
 * process that got there first could serve a pipe of this name without
 * holding the key, and this client would connect to it before the handshake
 * refused *it*. What that attacker gets is the client's own `hello`, which
 * carries a proof bound to this endpoint and expiring in a minute, and
 * nothing else: the core sends no session data until `welcome` is accepted.
 */

export type LinkEvent =
  | { type: "snapshot"; generation: number; payload: Buffer }
  | { type: "snapshotEnd"; generation: number }
  | { type: "output"; generation: number; payload: Buffer }
  | { type: "gap"; generation: number; missing: number }
  | {
      type: "exit";
      sessionKey: string;
      generation: number;
      exitCode: number | null;
    }
  | { type: "stale"; sessionKey: string; generation: number; current: number }
  | { type: "warning"; sessionKey: string; message: string }
  | { type: "bye"; reason: string; drain: boolean }
  | { type: "closed" };

export type EventSink = (event: LinkEvent) => void;

interface Pending {
  readonly resolve: (message: HostMessage) => void;
  readonly reject: (error: Error) => void;
  readonly timer: NodeJS.Timeout;
}

export class Link {
  private readonly pending = new Map<number, Pending>();
  private readonly decoder = new FrameDecoder();
  private tracker: OutputTracker | undefined;
  private welcome: ((message: HostMessage) => void) | undefined;
  private closed = false;

  private constructor(
    private readonly socket: Socket,
    private readonly events: EventSink,
  ) {
    socket.on("data", (chunk: Buffer) => this.read(chunk));
    socket.on("error", () => this.end());
    socket.on("close", () => this.end());
  }

  static connect(endpoint: string, events: EventSink): Promise<Link> {
    return new Promise((resolve, reject) => {
      const socket = connect(endpoint);
      const failed = (error: Error): void => {
        socket.destroy();
        reject(error);
      };
      socket.once("error", failed);
      socket.once("connect", () => {
        socket.off("error", failed);
        resolve(new Link(socket, events));
      });
    });
  }

  get alive(): boolean {
    return !this.closed;
  }

  /**
   * Closes this connection. For an attached connection that is a **detach**:
   * the host keeps the session, and only `destroy` ends one.
   */
  close(): void {
    this.end();
    this.socket.destroy();
  }

  /**
   * `hello`, and the proof that goes with it.
   *
   * `auth` is optional because the Rust host does not ask for one and would
   * ignore it; the TypeScript host refuses a connection without it. The caller
   * passes whatever the host it is talking to needs, and both are the same
   * protocol major — see {@link HelloAuth}.
   */
  async handshake(client: string, auth?: HelloAuth): Promise<Greeting> {
    const greeted = new Promise<HostMessage>((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error("the session host did not answer hello"));
      }, REQUEST_TIMEOUT_MS);
      timer.unref?.();
      this.welcome = (message) => {
        clearTimeout(timer);
        resolve(message);
      };
    });
    await this.notify({
      type: "hello",
      protocol: PROTOCOL_MAJOR,
      client,
      ...(auth === undefined ? {} : { auth }),
    });
    return acceptWelcome(await greeted, PROTOCOL_MAJOR);
  }

  /** Sends without waiting for an answer. */
  notify(message: ClientMessage): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.closed) {
        reject(new Error("the session host connection is closed"));
        return;
      }
      this.socket.write(jsonFrame(message), (error) =>
        error ? reject(error) : resolve(),
      );
    });
  }

  /** Sends and waits for the reply carrying `id`. */
  async request(id: number, message: ClientMessage): Promise<HostMessage> {
    const answer = new Promise<HostMessage>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error("the session host timed out"));
      }, REQUEST_TIMEOUT_MS);
      timer.unref?.();
      this.pending.set(id, { resolve, reject, timer });
    });
    try {
      await this.notify(message);
    } catch (error) {
      const waiter = this.pending.get(id);
      if (waiter !== undefined) {
        clearTimeout(waiter.timer);
        this.pending.delete(id);
      }
      throw error;
    }
    return answer;
  }

  /**
   * Starts checking output frames against `generation`. Called **before** the
   * `attach` request goes out, so no frame can arrive unchecked.
   */
  expectOutput(generation: number): void {
    this.tracker = new OutputTracker(generation);
  }

  private read(chunk: Buffer): void {
    this.decoder.push(chunk);
    for (;;) {
      let frame: Frame | undefined;
      try {
        frame = this.decoder.next();
      } catch {
        // The stream is not this protocol any more. There is no framing left
        // to resynchronise to, so the connection goes rather than the bytes
        // being guessed at.
        this.close();
        return;
      }
      if (frame === undefined) return;
      this.route(frame);
    }
  }

  private route(frame: Frame): void {
    if (frame.kind === "snapshot") {
      // A snapshot is the past being redrawn. It carries no sequence and must
      // not be checked against the live stream's numbering.
      this.events({
        type: "snapshot",
        generation: frame.generation,
        payload: frame.payload,
      });
      return;
    }
    if (frame.kind === "snapshotEnd") {
      this.events({ type: "snapshotEnd", generation: frame.generation });
      return;
    }
    if (frame.kind === "output") {
      const tracker = this.tracker;
      if (tracker === undefined) return;
      const delivery = tracker.observe(frame.generation, frame.sequence);
      if (delivery.kind === "write") {
        this.events({
          type: "output",
          generation: frame.generation,
          payload: frame.payload,
        });
        return;
      }
      if (delivery.kind === "gap") {
        this.tracker = undefined;
        this.events({
          type: "gap",
          generation: frame.generation,
          missing: delivery.missing,
        });
      }
      return;
    }
    let message: HostMessage;
    try {
      message = JSON.parse(frame.payload.toString("utf8")) as HostMessage;
    } catch {
      return;
    }
    switch (message.type) {
      case "welcome": {
        const waiter = this.welcome;
        this.welcome = undefined;
        waiter?.(message);
        return;
      }
      case "ok":
      case "error": {
        // A refusal of the handshake itself carries `id: 0` — no request had
        // been issued yet, so there is nothing to match it to by id, and the
        // one thing owed an answer is the outstanding `hello`. Without this,
        // "the host said no" arrives as "the host never answered", which is a
        // different problem with a different fix.
        if (
          message.type === "error" &&
          message.id === 0 &&
          this.welcome !== undefined
        ) {
          const greeter = this.welcome;
          this.welcome = undefined;
          greeter(message);
          return;
        }
        const waiter = this.pending.get(message.id);
        if (waiter === undefined) return;
        clearTimeout(waiter.timer);
        this.pending.delete(message.id);
        waiter.resolve(message);
        return;
      }
      case "exit":
        this.events({
          type: "exit",
          sessionKey: message.sessionKey,
          generation: message.generation,
          exitCode: message.exitCode,
        });
        return;
      case "stale":
        this.tracker = undefined;
        this.events({
          type: "stale",
          sessionKey: message.sessionKey,
          generation: message.generation,
          current: message.current,
        });
        return;
      case "warning":
        this.events({
          type: "warning",
          sessionKey: message.sessionKey,
          message: message.message,
        });
        return;
      case "bye":
        this.events({
          type: "bye",
          reason: message.reason,
          drain: message.drain,
        });
    }
  }

  private end(): void {
    if (this.closed) return;
    this.closed = true;
    // Waking every waiter is what keeps a dead host from turning into a set of
    // promises nobody ever settles.
    for (const [, waiter] of this.pending) {
      clearTimeout(waiter.timer);
      waiter.reject(new Error("the session host closed before answering"));
    }
    this.pending.clear();
    this.events({ type: "closed" });
  }
}

/* -------------------------------- discovery -------------------------------- */

/**
 * This user's SID.
 *
 * `whoami /user` is the only way to it without a native module, and its output
 * is stable across every supported Windows version. A failure here is fatal
 * rather than defaulted: the SID is half the pipe name, and guessing it would
 * aim this core at somebody else's host.
 */
export function currentSid(): string {
  const output = execFileSync("whoami", ["/user", "/fo", "csv", "/nh"], {
    encoding: "utf8",
    windowsHide: true,
  });
  // `"DOMAIN\user","S-1-5-21-…"`
  const sid = /"(S-1-[0-9-]+)"/.exec(output)?.[1];
  if (sid === undefined) {
    throw new Error(`could not read this user's SID from: ${output.trim()}`);
  }
  return sid;
}

/** The pipe this user's host of this data directory listens on. */
export function endpointFor(dataDir: string): string {
  return pipeEndpoint(currentSid(), dataDir);
}

/**
 * Which host this core starts.
 *
 * `ts` is the default from R6d: the TypeScript daemon under
 * `apps/desktop/src/session-host/`, bundled to `out/session-host/host.cjs`.
 * `rust` is the `armadra-session-host.exe` of `crates/session-host`, kept
 * reachable until R7 deletes the crate so that a Windows machine which hits a
 * problem with the new one has somewhere to stand.
 *
 * Both speak the same wire (`protocol.ts`); the only difference at this level
 * is what gets executed and whether a handshake proof is required.
 */
export type HostFlavour = "ts" | "rust";

export function hostFlavour(
  ambient: NodeJS.ProcessEnv = process.env,
): HostFlavour {
  return ambient.ARMADRA_SESSION_HOST === "rust" ? "rust" : "ts";
}

/** The file name of the TypeScript host's bundle. */
export const HOST_BUNDLE = "host.cjs";

/**
 * The Rust host binary, beside this executable — or wherever
 * `ARMADRA_SESSION_HOST_BIN` points, which is how a development run and the
 * probe reach one that is not next to Electron.
 */
export function resolveHostBinary(
  ambient: NodeJS.ProcessEnv = process.env,
): string {
  const named = ambient.ARMADRA_SESSION_HOST_BIN;
  if (named !== undefined && named !== "") {
    if (existsSync(named)) return named;
    throw new Error(
      `ARMADRA_SESSION_HOST_BIN does not point at a file: ${named}`,
    );
  }
  const sibling = join(dirname(process.execPath), HOST_BINARY);
  if (existsSync(sibling)) return sibling;
  throw new Error(`could not find ${HOST_BINARY} next to ${process.execPath}`);
}

/**
 * The places a built `host.cjs` can be, in the order they are tried.
 *
 * The same three the hook client's bundle uses, for the same reasons: the
 * packaged app's `resources/`, a sibling of the executable, and the built
 * tree (`out/core/main.js` → `out/session-host/host.cjs`) that a development
 * run works out of.
 */
export function hostBundleCandidates(
  ambient: NodeJS.ProcessEnv = process.env,
): string[] {
  const named = ambient.ARMADRA_SESSION_HOST_BUNDLE;
  if (named !== undefined && named !== "") return [named];
  const candidates: string[] = [];
  if (process.resourcesPath !== undefined) {
    candidates.push(join(process.resourcesPath, "session-host", HOST_BUNDLE));
  }
  candidates.push(join(dirname(process.execPath), "session-host", HOST_BUNDLE));
  candidates.push(join(__dirname, "..", "session-host", HOST_BUNDLE));
  return candidates;
}

export function resolveHostBundle(
  ambient: NodeJS.ProcessEnv = process.env,
): string {
  const candidates = hostBundleCandidates(ambient);
  const found = candidates.find((candidate) => existsSync(candidate));
  if (found !== undefined) return found;
  throw new Error(
    `could not find ${HOST_BUNDLE}; looked in:\n${candidates.map((c) => `  - ${c}`).join("\n")}`,
  );
}

export interface HostLaunch {
  readonly command: string;
  readonly args: readonly string[];
  readonly env: NodeJS.ProcessEnv;
}

/**
 * What to execute to bring a host up, for the flavour in effect.
 *
 * The TypeScript host is JavaScript, and a packaged machine is not guaranteed
 * to have a system `node` — so it is run by **this process' own executable**
 * with `ELECTRON_RUN_AS_NODE=1`, exactly as `armadra-hook`'s launcher does.
 * `ARMADRA_SESSION_HOST_RUNNER` overrides the interpreter, which is how a
 * plain-Node server shell and the integration tests reach it.
 *
 * Only the data directory is on the command line. Everything else the host
 * needs — the pipe name, the key, the lock — it derives from that one path,
 * because a command line is world-readable.
 */
export function hostLaunch(
  dataDir: string,
  ambient: NodeJS.ProcessEnv = process.env,
): HostLaunch {
  if (hostFlavour(ambient) === "rust") {
    return {
      command: resolveHostBinary(ambient),
      args: [dataDir],
      env: { ...ambient },
    };
  }
  const runner = ambient.ARMADRA_SESSION_HOST_RUNNER;
  return {
    command: runner !== undefined && runner !== "" ? runner : process.execPath,
    args: [resolveHostBundle(ambient), dataDir],
    env: { ...ambient, ELECTRON_RUN_AS_NODE: "1" },
  };
}

/**
 * Starts the host detached.
 *
 * Detached matters: a host started as an ordinary child would share this
 * process' console and process group, and the design is explicit that the
 * process owning the consoles must not be tied to one that exits.
 */
export function startHost(launch: HostLaunch): void {
  // CREATE_NEW_PROCESS_GROUP | DETACHED_PROCESS | CREATE_NO_WINDOW
  const child = spawn(launch.command, [...launch.args], {
    detached: true,
    stdio: "ignore",
    windowsHide: true,
    env: launch.env,
  });
  child.unref();
}

/** Connects, retrying with backoff until `within` elapses. */
export async function waitForHost(
  endpoint: string,
  events: EventSink,
  within: number = START_TIMEOUT_MS,
): Promise<Link> {
  const deadline = Date.now() + within;
  let wait = 50;
  for (;;) {
    try {
      return await Link.connect(endpoint, events);
    } catch (error) {
      if (Date.now() >= deadline) throw error;
      await new Promise((resolve) => setTimeout(resolve, wait));
      wait = Math.min(wait * 2, 500);
    }
  }
}
